R/kernel.r

```diff
@@ -242,6 +243,92 @@ complete = function(request) {
         cursor_end = start_position + nchar(c.info$token)))
 },
 
+inspect = function(request) {
+    make_section_header_plain <- function (txt) paste0("# ", txt, ":\n")
+    make_section_header_html <- function (txt) paste0("<h1>", txt, ":</h1>\n")
+    # 5.0 protocol:
+    code <- request$content$code
+    cursor_pos <- request$content$cursor_pos
+
+    # Get token under the cursor_pos.
+    # There may be better ways to do that than this.
+    token <- ""
+    for (i in seq(cursor_pos, nchar(code))) {
+        topic_candidate <- utils:::.guessTokenFromLine(code, i)
+        if (!grepl(token, topic_candidate)) break
+        token <- topic_candidate
+    }
+
+    if (nchar(token) == 0) {
+        data <- list()
+        found <- FALSE
+    } else {
+        data <- list()
+
+        class_data <- tryCatch(
+            eval(parse(text = paste0("class(", token, ")"))),
+            error = function (e) character(0))
+        if (length(class_data) != 0) {
+            section_title <- "Class attribute"
+            body <- paste(class_data, collapse = " ")
+            data[["text/plain"]] <- paste0(
+                make_section_header_plain(section_title),
+                body,
+                "\n\n")
+            data[["text/html"]] <- paste0(
+                make_section_header_html(section_title),
+                body,
+                "\n")
+        }
+
+        code_to_get_print_mimebundle <- paste0("IRdisplay::prepare_mimebundle(", token, ")")
+        print_data <- tryCatch(
+            eval(parse(text = code_to_get_print_mimebundle))$data,
+            error = function (e) list())
+        if (length(print_data) != 0) {
+            section_title <- "Printed form"
+            data[["text/plain"]] <- paste0(
+                data[["text/plain"]],
+                make_section_header_plain(section_title),
+                print_data[["text/plain"]],
+                "\n\n")
+            data[["text/html"]] <- paste0(
+                data[["text/html"]],
+                make_section_header_html(section_title),
+                print_data[["text/html"]],
+                "\n")
+        }
+
+        code_to_get_help_mimebundle <- paste0("IRdisplay::prepare_mimebundle(?", token, ")")
+        help_data <- tryCatch({
```

Comment:
`help_data` and `print_data` are created using almost identical code.

if there’s still more than 3 lines left after applying all my other ideas, make a function.